WebApp: add a profile for provisioning the server

WolfgangHG opened the pull request:
Part two to resolve #145: I added a profile "arq-provisioned" to the webapp archetype that provisions a server and runs the arquillian tests using this server.

This pull request is a request for comments ;-).

The provisioning profile required a change to the way the arquillian test deployment is created: as the provisioned server already contains the war file, the test deployment war file must have a different name - I added suffix "Tests", e.g. "MyApplicationTests.war".

Probably, it would be more reasonable to name the deployed app "Root.war" in case of provisioning, but this would probably conflict with the old school use case to develop an app without knowing how it will be deployed. So I would prefer to keep the name as is - I added hints to the end user "Readme.txt" to change the deployment name depending on the use case.

WolfgangHG committed:
WebApp: add a profile for provisioning the server

jfdenise commented:
@WolfgangHG , I realize that we should use in the arq profile the `<skipDeployment>true</skipDeployment>` config item of the WildFly Maven Plugin. So it will not conflict with the useless app deployment.

jfdenise commented:
I am not sure of the best, I wanted to highlight this skipDeployment in case it would make things simpler.

WolfgangHG commented:
I can't decide either. "skipDeployment" sounds better, as the arquillian deployment could keep its name. But it means more profiles. I leave it to the reviewers of this issue.

jmesnil commented:
I'm not particularly keen on having our archetypes propose multiple ways (aka profiles) to achieve the same thing.
This makes them more complex for the users that want to get started (which is the primary target for these archetypes).

So I would prefer that we provide a single way to compile/build and test a Web deployment on WildFly.

I think that the provisioned way is the better approach for development: you control the WildFly version from your Maven POM and you can test the *actual* deployment. There is no need to use ShrinkWrap to deploy a "fake" test deployment, the WAR is already there ready to be tested.

wdyt?

WolfgangHG commented:
@jmesnil I am not sure whether we can remove the shrinkwrap WAR creation, as we have to enrich the deployed archive with the server side test classes so that the [Arquillian ServletTestRunner](https://github.com/arquillian/arquillian-jakarta/blob/main/protocols/servlet-jakarta/src/main/java/org/jboss/arquillian/protocol/servlet5/runner/ServletTestRunner.java) can invoke them on the server side. They will not be part of the provisioned deployment.

Personally, I consider the archetype as a set of possibilities - it shows serveral ways to use WildFly, as it is currently doing for managed and remote instances. Using a provisioned server is another way. It is hard to collect them all in one project.
Would it be reasonable to add another archetype that focuses on provisioning and does not contain the arquillian-managed and arquillian-remote profiles? I agree with your concern that my current pull request makes the archetype project harder to understand for a beginner.

We could also remove the non-provisioning stuff and just point to it in "Readme.txt". This would follow your suggestion to keep it simple.

WolfgangHG commented:
Hmmm, just an idea: we could add some properties to the archetype which query from the user the needed functionality (e.g. "provisioning", "arquillian", "jakarta_faces" with "yes/no" values). 
In the pom.xml, conditional snippets using the Velocity Template Language seem to be possible: https://stackoverflow.com/questions/38831140/maven-archetype-optional-property-empty-by-default/60748225#60748225 so that e.g. profiles could be removed if "provisioning" is not selected.
Additionally, a groovy script could tailor the project created from the archetype to fit the requirements: https://ngeor.com/2019/09/07/optional-files-maven-archetype.html (e.g. remove the arquillian integration test sample file "SampleIT.java" or "faces-config.xml" if the features are not selected). 

But those archetype properties don't allow to add a description, so they should have meaningful names, and there might exist combinations that don't make sense.

WolfgangHG commented:
I split provisioning in two profiles (similar to the changes in the pull request draft for the ear archetype):

1. Profile "provision" just builds the server, without running arquillian tests. Here the web application is added to the deployment.
2. Profile "arq-provisioned" provisions a server (in a different directory), but does not add the web application to the deployment. So there is need to rename the arquillian deployment as it does not conflict with an existing deployment.

I think this makes the combination of features a bit more consistent. The only thing that is not perfect is that a provisioned server would probably contain a "ROOT.war" deployment. But I added comments to "Readme.txt" how to fix this.

jfdenise commented on wildfly-jakartaee-webapp-archetype/src/main/resources/archetype-resources/README.txt:
```diff
@@ -15,21 +15,57 @@ a persistence unit "${rootArtifactId}PersistenceUnit" which uses the JakartaEE d
+The project defines two profiles "provision" and "arq-provisioned" that create a provisioned WildFly server (in "target/server" or "target/server_arquillian")
+by auto discovering features necessary to run this application.
+
+The profile "provision" is meant to build to build a server that contains this web application as deployment.
```
to build is duplicated.

WolfgangHG replied:
Done.

jfdenise commented on wildfly-jakartaee-webapp-archetype/src/main/resources/archetype-resources/README.txt:
```diff
@@ -15,21 +15,57 @@ a persistence unit "${rootArtifactId}PersistenceUnit" which uses the JakartaEE d
+by auto discovering features necessary to run this application.
+
+The profile "provision" is meant to build to build a server that contains this web application as deployment.
+The profile "arq-provisioned" also builds a server, then runs t he arquillian tests. This profile does not add the deployment, so that the arquillian
```
white space after t

WolfgangHG replied:
Fixed.

jfdenise commented on wildfly-jakartaee-webapp-archetype/src/main/resources/archetype-resources/README.txt:
```diff
@@ -15,21 +15,57 @@ a persistence unit "${rootArtifactId}PersistenceUnit" which uses the JakartaEE d
+tests can deploy the app themselves enriched with additional test classes.
+
+As "persistence.xml" uses the datasource "java:comp/DefaultDataSource", Glow must be configured to create 
+the datasource in the WildFly server config. Currently, the add-on "create default H2 database" is configured:
```
The sentence "Currently, the add-on "create default H2 database" is configured" seems strange. "the default H2 datasource" perhaps?

WolfgangHG replied:
I applied your suggestion.

WolfgangHG commented:
After latest rebase, build failed due to the JBoss repository not being available. Had to do a `git commit --amend` to re-run the failed tests.

jmesnil commented:
We discussed this PR during the [WildFly Community Call](https://docs.google.com/document/d/1kjaMGFGtL3MPY58UCJMojLlTzl335DUm1L-P9lYk2HY/edit?tab=t.0#heading=h.y5aixk62m307) and agreed to move on with a single provisioned server.

* The webapp will provision WildFly and add the deployment
* The Arquillian test will be run against this provisioned server
  * it can test against the actual deployment (with `@RunAsClient`)
  * it can test against any classes by creating a Deployment with Shrinkwrap

We could replicate what the [getting-started archetyp](https://github.com/wildfly/wildfly-archetypes/tree/main/wildfly-getting-started-archetype/src/main/resources/archetype-resources/src/test/java) server is doing by having these 2 kind of tests

WolfgangHG commented:
@jmesnil Sorry, I lost track of this issue. Could you comment on my last question about the fundamental concept of the changes - if we make changes at all ;-)?
